src/ptvsd/adapter/ide.py

```diff
@@ -0,0 +1,290 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See LICENSE in the project root
+# for license information.
+
+from __future__ import absolute_import, print_function, unicode_literals
+
+import platform
+
+import ptvsd
+from ptvsd.common import json, log, messaging
+from ptvsd.common.compat import unicode
+from ptvsd.adapter import components
+
+
+class IDE(components.Component):
+    """Handles the IDE side of a debug session."""
+
+    message_handler = components.Component.message_handler
+
+    class Capabilities(components.Capabilities):
+        PROPERTIES = {
+            "supportsVariableType": False,
+            "supportsVariablePaging": False,
+            "supportsRunInTerminalRequest": False,
+            "supportsMemoryReferences": False,
+        }
+
+    class Expectations(components.Capabilities):
+        PROPERTIES = {
+            "locale": "en-US",
+            "linesStartAt1": True,
+            "columnsStartAt1": True,
+            "pathFormat": json.enum("path"),  # we don't support "uri"
+        }
+
+    def __init__(self, session, stream):
+        super(IDE, self).__init__(session, stream)
+
+        self.client_id = None
+        """ID of the connecting client. This can be 'test' while running tests."""
+
+        self._initialize_request = None
+        """The "initialize" request as received from the IDE, to propagate to the
+        server later."""
+
+        self._deferred_events = []
+        """Deferred events from the launcher and the server that must be propagated
+        only if and when the "launch" or "attach" response is sent.
+        """
+
+        assert not session.ide
+        session.ide = self
+
+        self.channel.send_event(
+            "output",
+            {
+                "category": "telemetry",
+                "output": "ptvsd.adapter",
+                "data": {"version": ptvsd.__version__},
+            },
+        )
+
+    def propagate_after_start(self, event):
+        # pydevd starts sending events as soon as we connect, but the IDE doesn't
+        # expect to see any until it receives the response to "launch" or "attach"
+        # request. If IDE is not ready yet, save the event instead of propagating
+        # it immediately.
+        if self._deferred_events is not None:
+            self._deferred_events.append(event)
+            log.debug("Propagation deferred.")
```

Comment:
>log.debug("Propagation deferred.") [](start = 12, length = 34)

Should this have `event.describe()` ?

Reply:
It had initially, but it's redundant, since `@message_handler` will put that on top of _every_ log entry coming from within the message handler. Looks like this:
```
D+00000.890: [handling #4 event "process" from Server-1]
             Propagation deferred.

D+00000.890: [handling #3 request "configurationDone" from IDE-1]
             Server-1 <-- {
                 "seq": 4,
                 "type": "request",
                 "command": "configurationDone"
             }
```